sys/kern/signal.c

```diff
@@ -201,6 +204,88 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
   return ERESTARTNOHAND;
 }
 
+int do_sigtimedwait(proc_t *p, sigset_t waitset, ksiginfo_t *kinfo,
+                    timespec_t *tsp) {
+  int timevalid = 0, error = 0, timeout = 0;
+  sigset_t saved_mask;
+  signo_t sig;
+  thread_t *td = p->p_thread;
+
+  if (tsp != NULL) {
+    if (tsp->tv_nsec < 0 || tsp->tv_nsec >= 1000000000) {
+      return EINVAL;
+    }
+    /* This is the same behaviour as in OpenBSD. */
+    if (tsp->tv_sec < 0) {
+      return EAGAIN;
+    }
+
+    timevalid = 1;
+    timeout = ts2hz(tsp);
+    /* Timeout set to 0 means to not block, but for sleepq it means to
+     * wait indefinitely */
+    if (timeout == 0) {
+      if (tsp->tv_nsec == 0 && tsp->tv_sec == 0) {
+        timeout = -1;
```

Comment:
I also think that if you move the comment that means "do not block" here and remove the comment in lines 225-226 then it would be more clear.

Reply:
Eventually provide some comment in style that from NetBSD (https://mimiker.ii.uni.wroc.pl/source/xref/NetBSD/sys/kern/sys_sig.c?r=53d22107#741). I really like that because it is immidiately clear what given values mean.